src/main/java/io/cryostat/net/web/http/AbstractAuthenticatedRequestHandler.java

```diff
@@ -164,16 +169,35 @@ private boolean isAuthFailure(ExecutionException e) {
 
     private void handleConnectionException(RoutingContext ctx, ConnectionException e) {
         Throwable cause = e.getCause();
-        if (cause instanceof SecurityException || cause instanceof SaslException) {
-            ctx.response().putHeader(JMX_AUTHENTICATE_HEADER, "Basic");
-            throw new HttpStatusException(427, "JMX Authentication Failure", e);
-        }
-        Throwable rootCause = ExceptionUtils.getRootCause(e);
-        if (rootCause instanceof ConnectIOException) {
-            throw new HttpStatusException(502, "Target SSL Untrusted", e);
-        }
-        if (rootCause instanceof UnknownHostException) {
-            throw new HttpStatusException(404, "Target Not Found", e);
+        try {
+            if (cause instanceof SecurityException || cause instanceof SaslException) {
+                ctx.response().putHeader(JMX_AUTHENTICATE_HEADER, "Basic");
+                throw new HttpStatusException(427, "JMX Authentication Failure", e);
+            }
+            Throwable rootCause = ExceptionUtils.getRootCause(e);
+            if (rootCause instanceof ConnectIOException) {
+                throw new HttpStatusException(502, "Target SSL Untrusted", e);
+            }
+            if (rootCause instanceof UnknownHostException) {
+                throw new HttpStatusException(404, "Target Not Found", e);
+            }
+        } finally {
+            this.removeCredentialsIfPresent(ctx);
         }
     }
+
+    private void removeCredentialsIfPresent(RoutingContext ctx) {
+        Optional<String> targetId = Optional.ofNullable(ctx.pathParam("targetId"));
+
+        targetId.ifPresent(
+                id -> {
+                    if (credentialsManager.getCredentials(id) != null) {
+                        try {
+                            credentialsManager.removeCredentials(id);
+                        } catch (IOException unused) {
+                            // handleConnectionException already throws an HTTPStatusException
```

Comment:
I think it'd be good to log this exception at least, even if it doesn't get propagated any further back the call stack.